lib/rest/lib/rest.c

```diff
@@ -0,0 +1,246 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <unistd.h>
+
+#include <bsd/string.h>
+#include <microhttpd.h>
+
+#include <rest/rest.h>
+
+#include <hse_util/assert.h>
+#include <hse_util/compiler.h>
+#include <hse_util/event_counter.h>
+#include <hse_util/hse_err.h>
+#include <hse_util/logging.h>
+#include <hse_util/table.h>
+
+struct server {
+    struct MHD_Daemon *daemon;
+    bool initialized;
+    int socket_fd;
+    char socket_path[sizeof(((struct sockaddr_un *)NULL)->sun_path)];
+};
+
+struct context {
+    struct MHD_Connection *connection;
+    enum rest_method method;
+};
+
+static struct server server;
+
+static struct table *route_map;
+
+merr_t
+rest_register_routes(size_t routec, const struct rest_route *routev)
+{
+    if (routec == 0)
+        return 0;
+
+    if (routec > 0 && !routev)
+        return merr(EINVAL);
+
+    for (size_t i = 0; i < routec; i++) {
+        const void *elem = table_append_object(route_map, (void *)(routev + i));
+        if (ev(!elem))
+            return merr(ENOMEM);
+    }
+
+    return 0;
+}
+
+static merr_t
+create_socket(const char *const socket_path, int *const socket_fd)
+{
+    struct sockaddr_un sock;
+    int tmp_fd;
+    int rc;
+    size_t n;
+
+    INVARIANT(socket_path);
+    INVARIANT(socket_fd);
+
+    /* In case the program exited inadvertently on the last run,
+     * remove the socket.
+     */
+    unlink(socket_path);
+
+    tmp_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (tmp_fd == -1)
+        return merr(errno);
+
+    sock.sun_family = AF_UNIX;
+    n = strlcpy(sock.sun_path, socket_path, sizeof(sock.sun_path));
+    if (n >= sizeof(sock.sun_path))
+        return merr(EINVAL);
+
+    rc = bind(tmp_fd, (const struct sockaddr *)&sock, sizeof(sock));
+    if (rc == -1) {
+        close(tmp_fd);
+
+        return merr(errno);
+    }
```

Comment:
close(2) can change errno, so capture the merr_t before calling it...
e.g.,

```
if (rc == -1) {
    merr_t err = merr(errno);

    close();
    return err;
}
```

Reply:
Hmmm, I can no longer find it...

Reply:
Looks like you might've been looking at an old commit or something. Not sure.

Reply:
It's hard to keep up with your commit frequency :-)

Reply:
I think you might've commented on code from the first commit in this PR. Note that there are two commits currently. Eventually they'll be squashed together.